packages/transaction-controller/src/hooks/SequentialPublishBatchHook.ts

```diff
@@ -0,0 +1,227 @@
+   */
+  getHook(): PublishBatchHook {
+    return this.#hook.bind(this);
+  }
+
+  async #hook({
+    from,
+    networkClientId,
+    transactions,
+  }: PublishBatchHookRequest): Promise<PublishBatchHookResult> {
+    log('Starting sequential publish batch hook', { from, networkClientId });
+
+    const pendingTransactionTracker =
+      this.#getPendingTransactionTrackerByChainId(networkClientId);
+    const results = [];
+
+    for (const transaction of transactions) {
+      try {
+        const transactionMeta = this.#getTransaction(String(transaction.id));
+
+        const transactionHash = await this.#publishTransaction(
+          this.#getEthQuery(networkClientId),
+          transactionMeta,
+        );
+        log('Transaction published', { transactionHash });
+
+        const transactionUpdated = {
+          ...transactionMeta,
+          hash: transactionHash,
+        };
+
+        const confirmationPromise = this.#waitForTransactionEvent(
+          pendingTransactionTracker,
+          transactionUpdated,
+          transactionHash,
+        );
+
+        pendingTransactionTracker.addTransactionToPoll(transactionUpdated);
+
+        await confirmationPromise;
+        results.push({ transactionHash });
+      } catch (error) {
+        log('Batch transaction failed', { transaction, error });
+        pendingTransactionTracker.stop();
+        throw rpcErrors.internal(`Failed to publish batch transaction`);
+      }
+    }
+    log('Sequential publish batch hook completed', { results });
+    pendingTransactionTracker.stop();
+
+    return { results };
+  }
+
+  /**
+   * Waits for a transaction event (confirmed, failed, or dropped) and resolves/rejects accordingly.
+   *
+   * @param pendingTransactionTracker - The tracker instance to subscribe to events.
+   * @param transactionMeta - The transaction metadata.
+   * @param transactionHash - The hash of the transaction.
+   * @returns A promise that resolves when the transaction is confirmed or rejects if it fails or is dropped.
+   */
+  async #waitForTransactionEvent(
+    pendingTransactionTracker: PendingTransactionTracker,
+    transactionMeta: TransactionMeta,
+    transactionHash: string,
+  ): Promise<TransactionMeta> {
+    return new Promise((resolve, reject) => {
+      const onConfirmed = this.#onConfirmedHandler({
+        transactionMeta,
+        transactionHash,
+        resolve,
+        pendingTransactionTracker,
+      });
+
+      const onFailedOrDropped = this.#onFailedOrDroppedHandler({
+        transactionMeta,
+        transactionHash,
+        reject,
+        pendingTransactionTracker,
+      });
+
+      this.#addListeners({
+        pendingTransactionTracker,
+        onConfirmed,
+        onFailedOrDropped,
+      });
+    });
+  }
+
+  #onConfirmedHandler({
+    transactionMeta,
+    transactionHash,
+    resolve,
+    pendingTransactionTracker,
+  }: OnConfirmedHandlerOptions): (txMeta: TransactionMeta) => void {
+    return (txMeta) => {
```

Comment:
Minor, could we use `.bind` to avoid returning a function?

Reply:
Sure, using `.bind` instead of passing a function.